src/cargo/core/resolver/mod.rs

```diff
@@ -1249,16 +1212,69 @@ fn activate_deps_loop(
+                            .filter_map(|(_, (ref new_dep, _, _))| {
+                                past_conflicting_activations.conflicting(&cx, &new_dep)
                             })
-                            .flat_map(|x| x)
-                            .find(|con| cx.is_conflicting(None, con))
+                            .next()
                         {
-                            conflicting_activations.extend(conflicting.clone());
+                            let mut conflicting = conflicting.clone();
+
+                            // If one of our deps conflicts with
+                            // then we will not succeed.
+                            // How ever if we are part of the reason that
+                            // one of our deps conflicts then
+                            // we can make a stronger statement
+                            // because we will definitely be activated when
+                            // we try our dep.
+                            conflicting.remove(&pid);
+
+                            conflicting_activations.extend(conflicting);
                             has_past_conflicting_dep = true;
                         }
                     }
+                    // If any of `remaining_deps` are known unresolvable with
+                    // us activated, then we extend our own set of
+                    // conflicting activations with theirs and its parent. We can do this
+                    // because the set of conflicts we found implies the
+                    // dependency can't be activated which implies that we
+                    // ourselves are incompatible with that dep, so we know that deps
+                    // parent conflict with us.
+                    if !has_past_conflicting_dep {
+                        if let Some(rel_deps) = past_conflicting_activations.get_dep_from_pid(&pid)
+                        {
+                            if let Some((other_parent, conflict)) = remaining_deps
+                                .iter()
+                                .flat_map(|other| other.flatten())
+                                // for deps related to us
+                                .filter(|&(_, ref other_dep)| rel_deps.contains(other_dep))
+                                .filter_map(|(other_parent, other_dep)| {
+                                    past_conflicting_activations
+                                        .filter_conflicting(
+                                            &cx,
+                                            &other_dep,
+                                            |con| con.contains_key(&pid)
+                                        )
+                                        .map(|con| (other_parent, con))
+                                })
+                                .next()
+                            {
+                                let mut conflict = conflict.clone();
```

Comment:
I think the `clone` here can be avoided, right? We could use `extend` with `filter` to remove `pid` and then later insert manually `other_parent` into `conflicting_activations`?

Reply:
That makes sense, but I can't get the iter to match correctly.

`conflicting_activations.extend(conflicting.iter().filter(|&(p, _)| p == &pid));`

gives:

```

error[E0271]: type mismatch resolving `<std::iter::Filter<std::collections::hash_map::Iter<'_, core::package_id::PackageId, core::resolver::ConflictReason>, [closure@src/cargo\core\resolver\mod.rs:1228:86: 1228:105 pid:_]> as std::iter::IntoIterator>::Item == (core::package_id::PackageId, core::resolver::ConflictReason)`
    --> src/cargo\core\resolver\mod.rs:1228:53
     |
1228 |                             conflicting_activations.extend(conflicting.iter().filter(|&(p, _)| p == &pid));
     |                                                     ^^^^^^ expected reference, found struct `core::package_id::PackageId`
     |
     = note: expected type `(&core::package_id::PackageId, &core::resolver::ConflictReason)`
                found type `(core::package_id::PackageId, core::resolver::ConflictReason)`
```

Reply:
I was reading the error message backwards. added a:
`.map(|(p, r)| (p.clone(), r.clone())` and it works.